client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param responseConsumer the {@link HttpAsyncResponseConsumer} callback
+     * @param responseListener the {@link ResponseListener} to notify when the request is completed or fails
+     * @param headers the optional request headers
+     */
+    public void performRequest(String method, String endpoint, Map<String, String> params,
+                               HttpEntity entity, HttpAsyncResponseConsumer<HttpResponse> responseConsumer,
+                               ResponseListener responseListener, Header... headers) {
         URI uri = buildUri(endpoint, params);
         HttpRequestBase request = createHttpRequest(method, uri, entity);
         setHeaders(request, headers);
-        //we apply a soft margin so that e.g. if a request took 59 seconds and timeout is set to 60 we don't do another attempt
-        long retryTimeoutMillis = Math.round(this.maxRetryTimeoutMillis / (float)100 * 98);
-        IOException lastSeenException = null;
+        FailureTrackingResponseListener failureTrackingResponseListener = new FailureTrackingResponseListener(responseListener);
         long startTime = System.nanoTime();
-        for (HttpHost host : nextHost()) {
-            if (lastSeenException != null) {
-                //in case we are retrying, check whether maxRetryTimeout has been reached, in which case an exception will be thrown
-                long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
-                long timeout = retryTimeoutMillis - timeElapsedMillis;
-                if (timeout <= 0) {
-                    IOException retryTimeoutException = new IOException(
-                            "request retries exceeded max retry timeout [" + retryTimeoutMillis + "]");
-                    retryTimeoutException.addSuppressed(lastSeenException);
-                    throw retryTimeoutException;
+        performRequest(startTime, nextHost().iterator(), request, responseConsumer, failureTrackingResponseListener);
+    }
+
+    private void performRequest(final long startTime, final Iterator<HttpHost> hosts, final HttpRequestBase request,
+                                final HttpAsyncResponseConsumer<HttpResponse> responseConsumer,
+                                final FailureTrackingResponseListener listener) {
+        final HttpHost host = hosts.next();
+        //we stream the request body if the entity allows for it
+        HttpAsyncRequestProducer requestProducer = HttpAsyncMethods.create(host, request);
+        client.execute(requestProducer, responseConsumer, new FutureCallback<HttpResponse>() {
+            @Override
+            public void completed(HttpResponse httpResponse) {
```

Comment:
sounds like we should rather expose the ability to provide a response consumer for the sync performRequest too. I don't have better ideas to make this work otherwise.